File: dotnet/Gitologist/Objects.cs
using Gitologist.Types;
using System.Text;

namespace Gitologist;

public static class Objects
{
    public static async Task<List<PackObject>> EnumerateObjects(string gitDir, string sha, HashSet<string>? visited = null)
    {
        var newVisited = visited ?? new HashSet<string>();
        if (newVisited.Contains(sha))
        {
            return new List<PackObject>();
        }
        newVisited.Add(sha);

        var objectData = await Utils.ReadObjectData(gitDir, sha);
        var nullIndex = Array.IndexOf(objectData, (byte)0);
        if (nullIndex == -1)
        {
            return new List<PackObject>();
        }

        var headerData = objectData[..nullIndex];
        var contentData = objectData[(nullIndex + 1)..];

        var header = Encoding.UTF8.GetString(headerData);
        var spaceIndex = header.IndexOf(' ');
        if (spaceIndex == -1)
        {
            return new List<PackObject>();
        }

        var typeString = header.Substring(0, spaceIndex);
        var type = typeString;

        var objects = new List<PackObject>
        {
            new PackObject
            {
                Type = type,
                Sha = sha,
                Content = contentData
            }
        };

        if (type == "commit")
        {
            var content = Encoding.UTF8.GetString(contentData);
            var lines = content.Split('\n');
            foreach (var line in lines)
            {
                if (line.StartsWith("parent "))
                {
                    var parentSha = line.Substring(7);
                    var parentObjects = await EnumerateObjects(gitDir, parentSha, newVisited);
                    objects.AddRange(parentObjects);
                }
                else if (line.StartsWith("tree "))
                {
                    var treeSha = line.Substring(5);
                    var treeObjects = await EnumerateObjects(gitDir, treeSha, newVisited);
                    objects.AddRange(treeObjects);
                }
            }
        }
        else if (type == "tree")
        {
            var entries = Utils.ParseTreeEntriesFromData(contentData);
            foreach (var entry in entries)
            {
                var entryObjects = await EnumerateObjects(gitDir, entry.Sha, newVisited);
                objects.AddRange(entryObjects);
            }
        }

        return objects;
    }

    public static async Task<List<PackObject>> GetAllObjects(string gitDir)
    {
        var objectsDir = Path.Combine(gitDir, "objects");
        var objects = new List<PackObject>();

        if (!Directory.Exists(objectsDir))
        {
            return objects;
        }

        var dirs = Directory.GetDirectories(objectsDir);

        foreach (var dir in dirs)
        {
            var dirName = Path.GetFileName(dir);
            if (dirName.Length != 2)
            {
                continue;
            }

            var files = Directory.GetFiles(dir);

            foreach (var file in files)
            {
                var fileName = Path.GetFileName(file);
                var sha = dirName + fileName;

                try
                {
                    var objectData = await Utils.ReadObject(gitDir, sha);
                    var headerEnd = objectData.IndexOf('\n');
                    if (headerEnd == -1)
                    {
                        continue;
                    }

                    var header = objectData.Substring(0, headerEnd);
                    var spaceIndex = header.IndexOf(' ');
                    if (spaceIndex == -1)
                    {
                        continue;
                    }

                    var type = header.Substring(0, spaceIndex);
                    var content = objectData.Substring(headerEnd + 1);

                    objects.Add(new PackObject
                    {
                        Type = type,
                        Sha = sha,
                        Content = Encoding.UTF8.GetBytes(content)
                    });
                }
                catch
                {
                    // Skip invalid objects
                }
            }
        }

        return objects;
    }
}

File: web/src/objects.ts
import { existsSync } from "node:fs";
import { join } from "node:path";

import type { PackObject } from "./packfile.ts";
import { parseTreeEntriesFromData, readObject, readObjectData } from "./utils.ts";

export async function enumerateObjects(
	gitDir: string,
	sha: string,
	visited: Set<string> = new Set(),
): Promise<PackObject[]> {
	if (visited.has(sha)) {
		return [];
	}
	visited.add(sha);

	const objectData = await readObjectData(gitDir, sha);
	const nullIndex = objectData.indexOf(0);

	if (nullIndex === -1) {
		return [];
	}

	const headerData = objectData.subarray(0, nullIndex);
	const contentData = objectData.subarray(nullIndex + 1);

	const header = headerData.toString("utf-8");
	const spaceIndex = header.indexOf(" ");

	if (spaceIndex === -1) {
		return [];
	}

	const typeString = header.slice(0, spaceIndex);
	const type = typeString as PackObject["type"];

	const objects: PackObject[] = [
		{
			type,
			sha,
			content: contentData,
		},
	];

	if (type === "commit") {
		const content = contentData.toString("utf-8");
		const lines = content.split("\n");
		for (const line of lines) {
			if (line.startsWith("parent ")) {
				const parentSha = line.slice(7);
				const parentObjects = await enumerateObjects(gitDir, parentSha, visited);
				objects.push(...parentObjects);
			} else if (line.startsWith("tree ")) {
				const treeSha = line.slice(5);
				const treeObjects = await enumerateObjects(gitDir, treeSha, visited);
				objects.push(...treeObjects);
			}
		}
	} else if (type === "tree") {
		const entries = parseTreeEntriesFromData(contentData);
		for (const entry of entries) {
			const entryObjects = await enumerateObjects(gitDir, entry.sha, visited);
			objects.push(...entryObjects);
		}
	}

	return objects;
}

export async function getAllObjects(gitDir: string): Promise<PackObject[]> {
	const objectsDir = join(gitDir, "objects");
	const objects: PackObject[] = [];

	if (!existsSync(objectsDir)) {
		return objects;
	}

	const { readdir } = await import("node:fs/promises");
	const dirs = await readdir(objectsDir);

	for (const dir of dirs) {
		if (dir.length !== 2) continue;

		const dirPath = join(objectsDir, dir);
		const files = await readdir(dirPath);

		for (const file of files) {
			const sha = dir + file;
			try {
				const objectData = await readObject(gitDir, sha);
				const headerEnd = objectData.indexOf("\n");
				const header = objectData.slice(0, headerEnd);
				const content = objectData.slice(headerEnd + 1);

				const spaceIndex = header.indexOf(" ");
				const type = header.slice(0, spaceIndex) as PackObject["type"];

				objects.push({
					type,
					sha,
					content: Buffer.from(content, "utf-8"),
				});
			} catch {
				// Skip invalid objects
			}
		}
	}

	return objects;
}
